Sofa/Component/Constraint/Lagrangian/Solver/src/sofa/component/constraint/lagrangian/solver/ConstraintForceExporter.h

```diff
@@ -0,0 +1,113 @@
+* FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License *
+* for more details.                                                           *
+*                                                                             *
+* You should have received a copy of the GNU Lesser General Public License    *
+* along with this program. If not, see <http://www.gnu.org/licenses/>.        *
+*******************************************************************************
+* Authors: The SOFA Team and external contributors (see Authors.txt)          *
+*                                                                             *
+* Contact information: contact@sofa-framework.org                             *
+******************************************************************************/
+#pragma once
+#include <sofa/component/constraint/lagrangian/solver/config.h>
+
+#include <sofa/component/constraint/lagrangian/solver/ConstraintSolverImpl.h>
+#include <sofa/core/behavior/MechanicalState.h>
+#include <sofa/core/behavior/SingleStateAccessor.h>
+
+#include <sofa/core/fwd.h>
+
+namespace sofa::component::constraint::lagrangian::solver
+{
+
+/**
+* Small utility component to compute forces (and wrenches in case of rigids)
+* from constraints computed by a constraint solver on a mechanical state,
+* at the end of each animation step.
+* Can also draw the forces for debug purpose (but it does not draw the wrenches if applicable).
+*/
+template <typename DataTypes>
+class ConstraintForceExporter : public sofa::core::behavior::SingleStateAccessor<DataTypes>
+{
+public:
+    SOFA_CLASS(ConstraintForceExporter, SOFA_TEMPLATE(sofa::core::behavior::SingleStateAccessor, DataTypes));
+
+    using Inherited = typename sofa::core::behavior::SingleStateAccessor<DataTypes>;
+
+    using Deriv = typename DataTypes::Deriv;
+    using VecDeriv = typename DataTypes::VecDeriv;
+
+protected:
+    ConstraintForceExporter();
+    ~ConstraintForceExporter() override = default;
+
+public:
+    void init() override;
+    void handleEvent(sofa::core::objectmodel::Event* event) override;
+    void draw(const sofa::core::visual::VisualParams* vparams) override;
+
+    Data<VecDeriv> d_constraintForces; ///< (output) Forces (or wrenches if rigids) computed from the mechanical state using the constraint solver.
+    Data<bool> d_draw; ///< (debug) draw forces (as an arrow) for each position of the mechanical object.
+    Data<double> d_drawForceScale; ///< (debug) Scale to apply on the force (draw as an arrow).
+
+    SingleLink<ConstraintForceExporter<DataTypes>, ConstraintSolverImpl, BaseLink::FLAG_STOREPATH | BaseLink::FLAG_STRONGLINK> l_constraintSolver; ///< Link to the constraint solver.
+
+    /// Check that DataTypes matches the MechanicalState.
+    template<class T>
+    static bool canCreate(T*& obj, core::objectmodel::BaseContext* context, core::objectmodel::BaseObjectDescription* arg)
```

Comment:
Is the canCreate function really needed ? Consistency checking with the rest of the scene may be done in init()

Reply:
I think the `canCreate` function is useful when the user does not specify the template in its scene. The `canCreate` function will be used to match the template of the mechanical object.

Reply:
Yes... but do we want to keep such an implicit mecanism ? And if so, isn't there any better option to implement it in a single place instead of implementing it/duplicating it in every component ?

Reply:
Yes, it was to do like C++ "type deduction", as for me setting the template (in scn) should be only mandatory for the mechanical state. Otherwise it feels like redundancy/duplicate information (and of course more clatter in the xml file...).
See Mapping, it is very rare you need to specify the templates, as it is 
 - bothersome (and hell to write when there is more than 1 template)
 - deductible
IMO, to check in canCreate() stage and to check at init() stage, is kind of like C++ compile-time assertion or a runtime assertion.

And yeah, when I was implementing the suggestion of @alxbilger to inherit from SingleStateAccessor, I was thinkiing it could be useful to do it direcly in StateAccessor (in a late PR, as it will totally not be the point of this PR)

Reply:
Speaking of user interface, I dislike the `template` attribute. Template refers to C++ code, but the term is not clear in a SOFA scene. Why would I use the same meaningless attribute to define the DoFs types in a mechanical state and for a matrix type in a linear solver? I would prefer something like `dataType`, or `dofType`, or `matrixType` (when it applies).

I agree that some components templates are complex to write. I see the situation similar to when we need to declare a C++ iterator. I prefer to write `auto it = vec.begin()` than the long iterator type. And again, I think `template` attribute is not clear for a scene designer perspective.

If there is a need to explicit things in a scene, I would prefer that a force field explicitly links to a mechanical state rather than defining a template. Finding implicitly the mechanical state in the current context is for me more confusing for the user compared to the automatic template deduction. Especially because `template` should not be exposed as it is.

Reply:
Agreed with @alxbilger, "template" is a very bad name if we focus on user that are not c++ developer. And there is actually there is complete section in Gui's bloopers's book (which i recommend reading) about how bad is leaking developer's terminology into the user interface and how developer like to do such a thing. 

@fredroy, the canCreate implemention puzzle me a lot. If there is a mstate attribute, it seems like to duplicate more or less what Link's internal type checking is doing and if there is none it do an implicit search mstate by do a query-path-string. Such behavior is not what most of of other sofa component are doing to "fill" links... in general it is done by using Node::get<MechanicalState>() in there init() function. Why doing such a thing in canCreate ?